### PR TITLE
Ensure that control points are at least the capture distance away from each other during campaign generation

### LINKED ISSUE #1912: "Could Not Find Front Line Point" Error
### Affected versions

Development build

### Development build

https://github.com/dcs-liberation/dcs_liberation/actions/runs/1647276730

### Description

Encountered this error at the end of Turn 2 of my updated Exercise Vegas Nerve campaign.
Capture

It may possibly be related to this: https://github.com/dcs-liberation/dcs_liberation/pull/1838

Note that while the Vegas Nerve campaign is an updated one (I have included the .miz and .yaml in the zip), those changes are on the southern end of the map (the addition of a FOB and an additional airfield). The section of frontline pathing (Tonopah to Groomlake) where this error occured has not changed since the version of Vegas Nerve in the release build.

### Save game and other files

[Files.zip](https://github.com/dcs-liberation/dcs_liberation/files/7814476/Files.zip)

### PR DESCRIPTION
**Pull request by MetalStormGhost**
Added a method in frontline.py which ensures the frontline conflict is always located between the control points and cannot move behind either one, where we couldn't find route segments. This is done by making sure the frontline is always placed at least one segment away from either control point.

Resolves #1912

### CODE REVIEW

**Comment by RndName** [comment]
Hey @DanAlbert can you take a look onto these changes? Not feeling confident enough to review them as i never ever touched some frontline things and this is the last fix one the list to 5.1.0 🚀

**Comment by DanAlbert** [comment]
> where we couldn't find route segments

Why would this happen? It seems like either this is the bug, or maybe `attack_distance` is wrong?

> This is done by making sure the frontline is always placed at least one segment away from either control point.

There's no guarantee that there _are_ segments. A single segment link is valid.

**Comment by MetalStormGhost** [comment]
Started to work on checking the minimum distance between control points during campaign generation. Discussion in Discord #dev https://discordapp.com/channels/595702951800995872/595706037449785364/929084322890268754

Now produces an error like this:
image

It's still currently checking all control points against all others, instead of just evaluating the ones which are immediately connected with supply routes. I was thinking we might only want to check the connected points. Comments?

File: game/campaignloader/mizcampaignloader.py [hunk 1]
```diff
@@ -283,6 +284,27 @@ def control_points(self) -> Dict[int, ControlPoint]:
                 control_point.captured_invert = fob.late_activation
                 control_points[control_point.id] = control_point
 
+            # Verify that all control points are at least the capture distance away from each other.
+            # If not, raise an exception to prevent the campaign generation.
+            for cp_index_a in control_points:
+                control_point_a = control_points[cp_index_a]
+                for cp_index_b in control_points:
+                    control_point_b = control_points[cp_index_b]
```
Comment (DanAlbert) [inline game/campaignloader/mizcampaignloader.py]: https://docs.python.org/3/library/itertools.html#itertools.product

```python
for a, b in itertools.product(control_points.values(), control_points.values()):
    ...
```

File: game/theater/frontline.py [hunk 2]
```diff
@@ -158,8 +158,11 @@ def point_from_a(self, distance: float) -> Point:
                 )
             else:
                 remaining_dist -= segment.attack_distance
-        raise RuntimeError(
-            f"Could not find front line point {distance} from {self.blue_cp}"
+        # Point was not found in any of the segment point_a:s. Returning the point from the last evaluated segment
+        # point_b, with heading towards the point_a, distance of (attack_distance - remaining_dist) meters.
+        return segment.point_b.point_from_heading(
+            segment.attack_heading.degrees - 180,
+            segment.attack_distance - remaining_dist,
```
Comment (DanAlbert) [inline game/theater/frontline.py]: This is unrelated to the PR description and so belongs in a separate PR.

Beyond that

> Point was not found in any of the segment point_a:s

Why not? Isn't that what the check above is suppose to guarantee?

**Review by DanAlbert — CHANGES_REQUESTED** [review]

File: game/theater/frontline.py (hunk 2 above)
Comment (MetalStormGhost) [reply game/theater/frontline.py]: No, that condition can be encountered e.g when the opposing force frontline units have been completely wiped out. In that situation _position_distance() will return the total attack_distance for that specific route, which is beyond the last point_a of the route (being actually at the point_b of the last segment).

File: game/campaignloader/mizcampaignloader.py (hunk 1 above)
Comment (MetalStormGhost) [reply game/campaignloader/mizcampaignloader.py]: Modified as suggested.

**Comment by RndName** [comment]
PR got stale, shall this be closed?

**Comment by MetalStormGhost** [comment]
> PR got stale, shall this be closed?

I think I had made both changes which Dan had requested and marked this as ready for review. Would you be willing to take a look at this?

**Comment by DanAlbert** [comment]
We're closing stale PRs. If this is still needed, please update and reopen.
